python/ray/dag/compiled_dag_node.py

```diff
@@ -510,139 +524,262 @@ def reset_and_wait_intermediate_future(self) -> Any:
+    #         overlap_gpu_communication: Whether to overlap GPU communication with
+    #             computation during DAG execution to improve performance.
+    #     Returns:
+    #         True if the next operation should not be executed; otherwise, False.
+    #     """
+    #     with self._recv_stream:
+    #         if self._read(overlap_gpu_communication):
+    #             return True
+    #     if self._compute(overlap_gpu_communication, class_handle):
+    #         return True
+    #     with self._send_stream:
+    #         if self._write(overlap_gpu_communication):
+    #             return True
+    #     return False
 
     def exec_operation(
         self,
         class_handle,
-        op_type: _DAGNodeOperationType,
         overlap_gpu_communication: bool = False,
     ) -> bool:
         """
         An ExecutableTask corresponds to a DAGNode. It consists of three
-        operations: READ, COMPUTE, and WRITE, which should be executed in
+        operations: read, compute, and write, which should be executed in
         order to ensure that each operation can read the correct intermediate
         result.
         Args:
             class_handle: The handle of the class to which the actor belongs.
-            op_type: The type of the operation. Possible types are READ,
-                COMPUTE, and WRITE.
             overlap_gpu_communication: Whether to overlap GPU communication with
                 computation during DAG execution to improve performance.
         Returns:
             True if the next operation should not be executed; otherwise, False.
         """
-        if op_type == _DAGNodeOperationType.READ:
-            with self._recv_stream:
-                return self._read(overlap_gpu_communication)
-        elif op_type == _DAGNodeOperationType.COMPUTE:
-            return self._compute(overlap_gpu_communication, class_handle)
-        elif op_type == _DAGNodeOperationType.WRITE:
-            with self._send_stream:
-                return self._write()
+        with self._recv_stream:
+            assert self._intermediate_future is None
+            try:
+                input_data = self.input_reader.read()
+                # When overlap_gpu_communication is enabled, wrap the result in
+                # a GPUFuture so that this read operation (communication) can
+                # be overlapped with computation.
+                self.wrap_and_set_intermediate_future(
+                    input_data, wrap_in_gpu_future=overlap_gpu_communication
+                )
+            except RayChannelError:
+                # Channel closed. Exit the loop.
+                return True
+
+        # To overlap GPU communication for NCCL recv, launch the NCCL recv operation,
+        # skip the normal compute operation, and return the future without waiting.
+        if not self.requires_nccl_read or not overlap_gpu_communication:
+            input_data = self.reset_and_wait_intermediate_future()
```

Comment:
Comment: Compute of `NcclRecvNode` is not needed.

Reply:
Comment: Compute of base `ClassMethodNode` and `NcclSendNode` needs to wait on the read future.